app/javascript/packages/document-capture/components/prompt-on-navigate.jsx

```diff
@@ -0,0 +1,20 @@
+import { useEffect } from 'react';
+
+/**
+ * While mounted, prompts the user to confirm navigation.
+ */
+function PromptOnNavigate() {
+  useEffect(() => {
+    function onBeforeUnload(event) {
+      event.preventDefault();
+      event.returnValue = '';
+    }
+
+    window.addEventListener('beforeunload', onBeforeUnload);
+    return () => window.removeEventListener('beforeunload', onBeforeUnload);
+  });
+
+  return null;
```

Comment:
don't we need to return the effect so that the callback can get passed around correctly?

Reply:
> don't we need to return the effect so that the callback can get passed around correctly?

If this were a custom hook, we might want to return something. I chose to implement it as a (non-visual) component instead. This made it easier to conditionally bind the prompt events, which would have been more difficult in hooks due to the [first rule of hooks around conditions](https://reactjs.org/docs/hooks-rules.html).

Basically, I'd wanted to take advantage of React's component lifecycle, without needing to show anything visual. Hence the `null` return value.